nixos/modules/services/web-servers/nginx/default.nix

```diff
@@ -673,23 +694,39 @@ in
       }
     ];
 
+    systemd.tmpfiles.rules = [
+      "d '${cfg.stateDir}' 0750 ${cfg.user} ${cfg.group} - -"
+      "d '${cfg.stateDir}/logs' 0750 ${cfg.user} ${cfg.group} - -"
+    ];
+
     systemd.services.nginx = {
       description = "Nginx Web Server";
       wantedBy = [ "multi-user.target" ];
       wants = concatLists (map (vhostConfig: ["acme-${vhostConfig.serverName}.service" "acme-selfsigned-${vhostConfig.serverName}.service"]) acmeEnabledVhosts);
       after = [ "network.target" ] ++ map (vhostConfig: "acme-selfsigned-${vhostConfig.serverName}.service") acmeEnabledVhosts;
       stopIfChanged = false;
-      preStart =
-        ''
+      preStart = ''
         ${cfg.preStart}
-        ${cfg.package}/bin/nginx -c ${configPath} -p ${cfg.stateDir} -t
-        '';
+        ${cfg.package}/bin/nginx -c '${configPath}' -p '${cfg.stateDir}' -t
+      '';
       serviceConfig = {
-        ExecStart = "${cfg.package}/bin/nginx -c ${configPath} -p ${cfg.stateDir}";
+        ExecStart = "${cfg.package}/bin/nginx -c '${configPath}' -p '${cfg.stateDir}'";
         ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
         Restart = "always";
         RestartSec = "10s";
         StartLimitInterval = "1min";
+        # User and group
+        User = cfg.user;
+        Group = cfg.group;
+        # Runtime directory and mode
+        RuntimeDirectory = "nginx";
+        RuntimeDirectoryMode = "0750";
+        # Capabilities
+        AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" ];
+      } // optionalAttrs (cfg.workerRlimitCore != null) {
+        LimitCORE = "${cfg.workerRlimitCore}";
```

Comment:
This will set this limit for the master, whereas the value is meant to be for each worker.

I don't think we should change the upstream option's semantics.

Therefore, I don't think we need the two new module options - people can just use `extraConfig`.

Reply:
This option need to crrect fix this warning:
`nginx[13580]: 2019/05/30 21:24:22 [alert] 13582#13582: setrlimit(RLIMIT_NOFILE, 102400) failed (1: Operation not permitted)`
Another method - add options 
`AmbientCapabilities = [ "CAP_SYS_RESOURCE" ];`
or
```
LimitCORE = "infinity";
LimitNOFILE = "infinity";
```

Reply:
The error message is coming from nginx failing to call `setrlimit` when creating worker processes - see [ngx_worker_process_init](https://github.com/nginx/nginx/blob/4bf4650f2f10f7bbacfe7a33da744f18951d416d/src/os/unix/ngx_process_cycle.c#L810)

Adding `AmbientCapabilities = [ "CAP_SYS_RESOURCE" ];` would probably the better way to fix this.

Reply:
Updated, thanks.